### ISSUE TITLE
`Sync.worker.js` import error Module not found: ESM packages in NextJs

### ISSUE THREAD

**Issue by SandroMaglione** [evoluhq/evolu#169]
I am getting the following error when using `useQuery` in NextJs:

<img width="971" alt="Screenshot 2023-06-14 at 12 33 56" src="https://github.com/evoluhq/evolu/assets/10065056/8263aea5-99d5-428b-bc4c-39c0ba650e53">

### NextJs
- `/app` directory (`page.tsx`)
- `"use client"`
- Version: `13.4.5`

```tsx
"use client";
import { useQuery } from "@/lib/schema/schema";

export default function Home() {
  const { rows } = useQuery(
    (db) => db.selectFrom("word").selectAll(),
    (row) => row
  );

  return (
    <main>
      <pre>{JSON.stringify(rows)}</pre>
    </main>
  );
}
```

### Evolu
- Version: `^6.1.1`

### Schema (`@effect/schema`)
- Version: `^0.20.1`

### Data (`@effect/data`)
- Version: `^0.12.6`

***

Any idea how to solve? Is this an internal ESM issue?

**Comment by steida** [comment]
Try to add `experimental: { esmExternals: "loose" }` in next.config.js. I think it's some Next.js bug.

**Comment by SandroMaglione** [comment]
> Try to add `experimental: { esmExternals: "loose" }` in next.config.js. I think it's some Next.js bug.

This works, but it seems to give some problems with reloads (the page does not update after changes and other issues with `setState`). Feel free to close if you think this is related to Next.

**Comment by steida** [comment]
@SandroMaglione Can you check https://github.com/evoluhq/evolu/tree/main/examples/nextjs, please?

**Comment by steida** [comment]
Related:

> When a dependency package has "type": "module" in package.json, and uses new Worker(new URL("./worker.js", import.meta.url)) to create a web worker, next build failed with

https://github.com/vercel/next.js/issues/33914

**Comment by steida** [comment]
If I remove `"type": "module"`, Next.js suddenly works without `experimental: { esmExternals: "loose" }`. The problem is it's not a valid ESM, so Node.js will refuse to work. If I change file extensions to `.mjs`, Next.js complains again.

**Comment by steida** [comment]
ESM will be supported in React Native 0.72. Electron will support it soon too https://github.com/electron/electron/pull/37535. I'm considering releasing evolu-cjs lib build from evolu for edge cases and old bundlers.

**Comment by steida** [comment]
So I used rollup with the rollup-plugin-off-main-thread plugin, and it worked. I was curious why so I compared codes, and the only difference was that the rollup output striped './' from the worker path. So I stripped it from my source code, and suddenly, Next.js works without `esmExternals: "loose"` 🤦

I spent weeks studying all Web Worker issues (e.g., https://github.com/lgarron/web-worker-compat). I was reading esbuild, rollup, and Next.js issues... for no real outcome because I used the wrong path format. And because Firefox 114 finally supports the module Web Worker.
